src/llmq/quorums_chainlocks.cpp

```diff
@@ -0,0 +1,424 @@
+{
+    if (!sporkManager.IsSporkActive(SPORK_19_CHAINLOCKS_ENABLED)) {
+        return;
+    }
+
+    CChainLockSig clsig;
+    {
+        LOCK(cs);
+
+        if (recoveredSig.id != lastSignedRequestId || recoveredSig.msgHash != lastSignedMsgHash) {
+            // this is not what we signed, so lets not create a CLSIG for it
+            return;
+        }
+        if (bestChainLock.nHeight >= lastSignedHeight) {
+            // already got the same or a better CLSIG through the CLSIG message
+            return;
+        }
+
+        clsig.nHeight = lastSignedHeight;
+        clsig.blockHash = lastSignedMsgHash;
+        clsig.sig = recoveredSig.sig;
+    }
+    ProcessNewChainLock(-1, clsig, ::SerializeHash(clsig));
+}
+
+void CChainLocksHandler::ScheduleInvalidateBlock(const CBlockIndex* pindex)
+{
+    // Calls to InvalidateBlock and ActivateBestChain might result in re-invocation of the UpdatedBlockTip and other
+    // signals, so we can't directly call it from signal handlers. We solve this by doing the call from the scheduler
+
+    scheduler->scheduleFromNow([this, pindex]() {
+        DoInvalidateBlock(pindex, true);
+    }, 0);
+}
+
+// WARNING, do not hold cs while calling this method as we'll otherwise run into a deadlock
+void CChainLocksHandler::DoInvalidateBlock(const CBlockIndex* pindex, bool activateBestChain)
+{
+    auto& params = Params();
+
+    {
+        LOCK(cs_main);
+
+        // get the non-const pointer
+        CBlockIndex* pindex2 = mapBlockIndex[pindex->GetBlockHash()];
+
+        CValidationState state;
+        if (!InvalidateBlock(state, params, pindex2)) {
+            LogPrintf("CChainLocksHandler::UpdatedBlockTip -- InvalidateBlock failed: %s\n", FormatStateMessage(state));
+            // This should not have happened and we are in a state were it's not safe to continue anymore
+            assert(false);
+            return;
+        }
+    }
+
+    CValidationState state;
+    if (activateBestChain && !ActivateBestChain(state, params)) {
+        LogPrintf("CChainLocksHandler::UpdatedBlockTip -- ActivateBestChain failed: %s\n", FormatStateMessage(state));
+        // This should not have happened and we are in a state were it's not safe to continue anymore
+        assert(false);
+        return;
+    }
+}
+
+bool CChainLocksHandler::HasChainLock(int nHeight, const uint256& blockHash)
+{
+    if (!sporkManager.IsSporkActive(SPORK_19_CHAINLOCKS_ENABLED)) {
+        return false;
+    }
+
+    LOCK(cs);
+    return InternalHasChainLock(nHeight, blockHash);
+}
+
+bool CChainLocksHandler::InternalHasChainLock(int nHeight, const uint256& blockHash)
+{
+    AssertLockHeld(cs);
+
+    if (!bestChainLockBlockIndex) {
+        return false;
+    }
+
+    if (nHeight > bestChainLockBlockIndex->nHeight) {
+        return false;
+    }
+
+    if (nHeight == bestChainLockBlockIndex->nHeight && blockHash == bestChainLockBlockIndex->GetBlockHash()) {
```

Comment:
`nHeight == bestChainLockBlockIndex->nHeight` should stop verification in any case i.e. this should be
```
if (nHeight == bestChainLockBlockIndex->nHeight) {
    return blockHash == bestChainLockBlockIndex->GetBlockHash();
}
````